src/platforms/web/runtime/directives/show.js

```diff
@@ -17,22 +17,24 @@ export default {
     if (value && transition && transition.appear && !isIE9) {
       enter(vnode)
     }
-    el.style.display = value ? '' : 'none'
+    const originalDisplay = el.style.display
+    el.style.display = value ? originalDisplay : 'none'
+    el.dataset.__vOriginalDisplay = originalDisplay
```

Comment:
I don't think IE9~10 supports `dataset`, so we may have to store the value directly on the element itself.